cylc/uiserver/websockets/tornado.py

```diff
@@ -58,19 +59,24 @@ async def _handle(self, ws, request_context):
         await self.on_open(connection_context)
         pending = set()
         while True:
+            message = None
             try:
                 if connection_context.closed:
                     raise ConnectionClosedException()
                 message = await connection_context.receive()
             except ConnectionClosedException:
                 break
+            except QueueEmpty:
+                pass
             finally:
                 if pending:
                     (_, pending) = await wait(pending, timeout=0, loop=self.loop)
 
-            task = ensure_future(
-                self.on_message(connection_context, message), loop=self.loop)
-            pending.add(task)
+            if message:
+                task = ensure_future(
+                    self.on_message(connection_context, message), loop=self.loop)
+                pending.add(task)
+            await sleep(1)
```

Comment:
Could probably only wait if no message was received:
```
[I 2020-05-10T09:38:48.469 JupyterHub log:174] 200 GET /hub/error/503?url=%2Fuser%2Fsutherlander%2Ffavicon.png (@127.0.0.1) 3.47ms
We have 6 tasks
We have 6 tasks
We have 6 tasks
09:39:15.051 [ConfigProxy] info: 200 GET /api/routes 
[I 2020-05-10T09:39:15.073 JupyterHub proxy:320] Checking routes
We have 6 tasks
We have 6 tasks
We have 8 tasks
We have 6 tasks
We have 6 tasks
We have 6 tasks
gen exit
gen exit
We have 8 tasks
We have 8 tasks
We have 8 tasks
We have 10 tasks
We have 8 tasks
We have 8 tasks
We have 8 tasks
We have 8 tasks
We have 8 tasks
We have 8 tasks
gen exit
We have 3 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
We have 1 tasks
^C[C 2020-05-10T09:43:24.323 JupyterHub app:2684] Received signal SIGINT, initiating shutdown...
[I 2020-05-10T09:43:24.324 JupyterHub app:2360] Cleaning up single-user servers...
[I 2020-05-10T09:43:24.325 JupyterHub proxy:706] Cleaning up proxy[14716]...
09:43:24.325 [ConfigProxy] warn: Terminated
2020-05-10 09:43:24,326 cylc.uiserver.main INFO     exiting...
2020-05-10 09:43:24,335 cylc.uiserver.main INFO     exit success
[I 2020-05-10T09:43:26.171 JupyterHub app:2392] ...done
(proto-uiserver) sutherlander@cortex-vbox:cylc-uiserver$ jupyterhub
```

Reply:
>Could probably only wait if no message was received: 

Great idea! Diff applied (thanks!)

>Is 1.0sec delay Ok? shorter? longer?

Good question. My first try had `sleep(0)`. I think one blog post I was looking at, with examples for non-blocking asyncio queue, had the sleep(1) so I went with that.

Just had a look at CPython, but looks like it simply [waits on a future](https://github.com/python/cpython/blob/2c3d508c5fabe40dac848fb9ae558069f0576879/Lib/asyncio/queues.py#L166), using CPython's interval default interval (couldn't find the source for that :).

Any thoughts on `0`, `1`, or greater anybody? Happy to update it.

Reply:
This fine I think, because the looping over our async gen (iterator) happens in a different loop;
https://github.com/cylc/cylc-uiserver/blob/14e9570aa0b7823ce89c93f9a822a8925a70cb01/cylc/uiserver/websockets/tornado.py#L191-L207
so 1sec (average of 0.5sec) to setup a new message on that connection.

Reply:
(I tested with a 0 second sleep and didn't see any CPU consequences)